A87-mtls-spiffe-support.md

```diff
@@ -0,0 +1,188 @@
+   configured trust map, reject the certificate.
+
+3. Verify the peer certificate using the default security library using
+   the SPIFFE trust bundle certificates as roots.
+
+SAN matching will be performed as configured via xDS, just as it is with
+X509 verification.
+
+gRPC currently supports only one certificate provider implementation,
+which is the `file_watcher` provider.  We will add a new field to the
+configuration of this provider implementation configuration called
+`spiffe_trust_bundle_map_file`.  This field will specify the path to
+the SPIFFE trust bundle map.
+
+If the `spiffe_trust_bundle_map_file` field is unset, then the
+`ca_certificate_file` field will be used exactly as it is today, and
+the certificate provider will return CA certificates.
+
+If the `spiffe_trust_bundle_map_file` field is set, the certificate
+provider will return the SPIFFE trust bundle map from the specified file.
+If the `ca_certificate_file` field is also set, it will be ignored,
+which is helpful for forward compatibility: bootstrap configs can start
+setting the new field regardless of what gRPC version is in use, which
+results in older versions returning CA certificates and newer versions
+returning the SPIFFE trust bundle map.
+
+When reading the SPIFFE trust bundle map file, the certificate provider
+will parse the JSON to ensure its validity and store the map in memory.
+If an error occurs during the initial load (e.g., a failure to read
+the file, or the file contains invalid entries), then there is no valid
+map, which means new connection attempts will fail the TLS handshake.
+For subsequent loads, errors will not affect the existing in-memory
+trust bundle map.  An empty map will be considered a valid (but empty)
+trust bundle map.
+
+Note that the [SPIFFE bundle format] and [Publishing SPIFFE bundle format]
+are partially supported:
+- Only the `keys` and `spiffe_sequence` elements of the JWK set are supported.
+- Only the `kty`, `use`, and `x5c` elements of `keys` are supported.
+- Instead of ignoring individual JWK entries in case of issues, we ignore
+  the whole trust bundle.
+
+### SPIFFE Certificate Verification in TlsCredentials
+
+The xDS mTLS support is built on top of existing mTLS functionality in
+TlsCredentials.  This section describes how the SPIFFE functionality
+will work in TlsCredentials and how the xDS functionality builds on top
+of it.  The details of this are different in each language.
+
+#### C++
+
+In C++, TlsCredentials natively supports a CertificateProvider API with
+essentially the same semantics as in xDS.  We will add the ability for
+the `FileWatcherCertificateProvider` to be instantiated with a SPIFFE
+trust bundle map file instead of a CA certificate file.
+
+TODO: @erm-g to fill in API details.  We'll probably need to introduce
+an options struct for FileWatcherCertificateProvider.
+
+#### Java
+
+In Java, we need new API to work with [SPIFFE] and [SPIFFE bundle
```

Comment:
This is _internal_ API, not public. This section documents how the xDS part is implemented, but it _doesn't_ describe any changes to TlsCredentials for use outside of xDS.

Reply:
Okay.  Do we not plan to support this functionality with TlsCreds in Java?

Reply:
@gtcooke94, we'll want to flesh this out a bit.

Reply:
How would you like to see this fleshed out with regards to the work Andrey did?